lib/Transforms/TensorToScalars/TensorToScalars.cpp

```diff
@@ -0,0 +1,149 @@
+                                                         TypeRange resultTypes,
+                                                         Value input,
+                                                         Location loc) {
+  // This pass only operates on tensors of static shape
+  RankedTensorType inputType = dyn_cast<RankedTensorType>(input.getType());
+  if (!inputType || !inputType.hasStaticShape()) return {};
+
+  // Create extract ops in "natural" order (dimension-by-dimension)
+  SmallVector<Value> values;
+  for (auto dim : inputType.getShape()) {
+    for (int i = 0; i < dim; ++i) {
+      Value index = builder.create<arith::ConstantIndexOp>(loc, i);
+      Value element = builder.create<tensor::ExtractOp>(loc, input, index);
+      values.push_back(element);
+    }
+  }
+  return values;
+}
+
+class ConvertFromElementsOp
+    : public OneToNOpConversionPattern<tensor::FromElementsOp> {
+ public:
+  using OneToNOpConversionPattern<
+      tensor::FromElementsOp>::OneToNOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor::FromElementsOp op, OpAdaptor adaptor,
+      OneToNPatternRewriter &rewriter) const override {
+    // OneToNConversion has no Conversion-level illegality handling
+    if (typeConverter->isLegal(op)) return failure();
+
+    // This pass only operates on tensors of static shape,
+    // but no check is necessary here as from_elements' shape is always static
+
+    // Replace the current op with the flattened operands.
+    // This should already match the "natural" order expected by this pass.
+    rewriter.replaceOp(op, adaptor.getFlatOperands(),
+                       adaptor.getResultMapping());
+    return success();
+  }
+};
+
+class ConvertInsertOp : public OneToNOpConversionPattern<tensor::InsertOp> {
+ public:
+  using OneToNOpConversionPattern<tensor::InsertOp>::OneToNOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      tensor::InsertOp op, OpAdaptor adaptor,
+      OneToNPatternRewriter &rewriter) const override {
+    // OneToNConversion has no Conversion-level illegality handling
+    if (typeConverter->isLegal(op)) return failure();
+
+    // This pass only operates on tensors of static shape
+    if (!op.getResult().getType().hasStaticShape()) return failure();
+
+    // Compute the insertion offset (in dimension-by-dimension order):
+    int64_t multiplier = 1;
+    int64_t offset = 0;
+    for (auto [dim, idx] :
+         llvm::zip(op.getResult().getType().getShape(), op.getIndices())) {
+      // We can only support statically known indices
+      // that have been constant-folded to a single arith.constant op
+      auto cidx = idx.getDefiningOp<arith::ConstantIndexOp>();
+      if (!cidx) return failure();
+      offset += cidx.value() * multiplier;
+      multiplier *= dim;
+    }
+
+    // get converted "tensor" operand from op (likely a unrealized_builtin_cast)
+    SmallVector<Value> elements = adaptor.getOperands()[1];
```

Comment:
nit: small readability thing, but you can grab `adaptor.getScalar()` and `adaptor.getDest()` instead of indexing into the operands

Reply:
Nice catch, thanks! I'm not sure why I didn't originally use them here

Reply:
Ah, because the function takes an `OpAdaptor` and not a `tensor::InsertOp::Adaptor`. Let me see if it works with that, too.

Reply:
Unfortunately, it does not - looks like this is a limitation of the 1toNConversion framework.

Reply:
argh! thank you for trying anyway, sorry about that! TIL